### ISSUE TITLE
Elassandra 5.5.0.24 not working with ec2snitch

### ISSUE THREAD

**Issue by DivsDibs** [strapdata/elassandra#236]
**Elassandra version**: **5.5.0.24**

**Plugins installed**: [None]

**JVM version** (`java -version`): **1.8.0_191**

**OS version** (`uname -a` if on a Unix-like system): **Linux ip-10-XX-YY-ZZZ.abcd..com 3.10.0-693.21.1.el7.x86_64 #1 SMP Fri Feb 23 18:54:16 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux**

**Problem Description**:
Elassandra Cluster on a Single AWS Region with the DC being spread over 3 AZ(s). We are using ec2snitch and facing the following exception while starting the 2nd node in the cluster. The 1st one comes up fine. 

**Steps to reproduce**:
 1. Launch ec2 instance, which only has pvt ipV4 address [no public ip address] 
 2. Download/Install the tar ball 
 3. `cassandra.yaml` is as follows:
```
     endpoint_snitch: Ec2Snitch
     listen_address: 10.XX.YY.ZZZ
     broadcast_address:
     rpc_address: 0.0.0.0
     broadcast_rpc_address: 10.XX.YY.ZZZ
seed_provider:
    - class_name: org.apache.cassandra.locator.SimpleSeedProvider
      parameters:
          - seeds: "10.X1.Y1.Z1,10.XX.YY.ZZZ"`

```
 3. Bring elassandra nodes up using `./cassandra -e`. The 1st node comes up; the 2nd & 3rd node fails with the below exception

```
2018-11-09 03:22:38,945 INFO  [main] org.elassandra.discovery.CassandraDiscovery.doStart(CassandraDiscovery.java:152) localNode name=10.XX.YY.ZZZ id=<<UUID>> localAddress=/10.XX.YY.ZZZ publish_host=10.XX.YY.ZZZ:1234
{5.5.0}: Initialization Failed ...
- NullPointerException[null]
2018-11-09 03:22:38,949 ERROR [main] org.apache.cassandra.service.ElassandraDaemon.main(ElassandraDaemon.java:588) Exception
java.lang.NullPointerException: null
        at org.elassandra.discovery.CassandraDiscovery.updateClusterGroupsFromGossiper(CassandraDiscovery.java:286)
        at org.elassandra.discovery.CassandraDiscovery.doStart(CassandraDiscovery.java:172)
        at org.elasticsearch.common.component.AbstractLifecycleComponent.start(AbstractLifecycleComponent.java:69)
        at org.elasticsearch.node.Node.start(Node.java:757)
        at org.apache.cassandra.service.ElassandraDaemon.activate(ElassandraDaemon.java:232)
        at org.apache.cassandra.service.ElassandraDaemon.main(ElassandraDaemon.java:551)`
```
Whereas in Node 1 which is already up it shows:
```
2018-11-09 03:22:31,167 ERROR [GossipStage:1] org.apache.cassandra.service.CassandraDaemon$2.uncaughtException(CassandraDaemon.java:231) Exception in thread Thread[GossipStage:1,5,main]
java.lang.NullPointerException: null
        at org.elassandra.discovery.CassandraDiscovery.updateNode(CassandraDiscovery.java:345)
        at org.elassandra.discovery.CassandraDiscovery.onChange(CassandraDiscovery.java:503)
        at org.apache.cassandra.gms.Gossiper.doOnChangeNotifications(Gossiper.java:1242)
        at org.apache.cassandra.gms.Gossiper.applyNewStates(Gossiper.java:1225)
        at org.apache.cassandra.gms.Gossiper.applyStateLocally(Gossiper.java:1168)
        at org.apache.cassandra.gms.GossipDigestAck2VerbHandler.doVerb(GossipDigestAck2VerbHandler.java:49)
        at org.apache.cassandra.net.MessageDeliveryTask.run(MessageDeliveryTask.java:66)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at org.apache.cassandra.concurrent.NamedThreadFactory.lambda$threadLocalDeallocator$0(NamedThreadFactory.java:81)
        at java.lang.Thread.run(Thread.java:748)
```
**RCA**
When we trace the code, it is seen that in both the bombed lines the erroneous code is
`
InetAddresses.forString(epState.getApplicationState(ApplicationState.INTERNAL_IP).value),
`

**Comment by vroyer** [comment]
Hi,
The 2 errors come because gossip INTERNAL_IP is null, probably because *broadcast_address* is set to nothing. You should comment out the *broadcast_address* to fall back to the *listen_address*.
Thanks.

**Comment by gouthamaenugu11** [comment]
Vroyer,
We tried commenting out Broadcast_address still it is the same issue. Any other way we can try?

**Comment by vroyer** [comment]
According to the code of the EC2 snitch, INTERNAL_IP can also be null because the awsApiCall(PRIVATE_IP_QUERY_URL) returned null in your AWS environment ?

```
public class Ec2MultiRegionSnitch extends Ec2Snitch
{
    private static final String PUBLIC_IP_QUERY_URL = "http://169.254.169.254/latest/meta-data/public-ipv4";
    private static final String PRIVATE_IP_QUERY_URL = "http://169.254.169.254/latest/meta-data/local-ipv4";
    private final String localPrivateAddress;

    public Ec2MultiRegionSnitch() throws IOException, ConfigurationException
    {
        super();
        InetAddress localPublicAddress = InetAddress.getByName(awsApiCall(PUBLIC_IP_QUERY_URL));
        logger.info("EC2Snitch using publicIP as identifier: {}", localPublicAddress);
        **localPrivateAddress = awsApiCall(PRIVATE_IP_QUERY_URL);**
        // use the Public IP to broadcast Address to other nodes.
        DatabaseDescriptor.setBroadcastAddress(localPublicAddress);
        if (DatabaseDescriptor.getBroadcastRpcAddress() == null)
        {
            logger.info("broadcast_rpc_address unset, broadcasting public IP as rpc_address: {}", localPublicAddress);
            DatabaseDescriptor.setBroadcastRpcAddress(localPublicAddress);
        }
    }

    @Override
    public void gossiperStarting()
    {
        super.gossiperStarting();
       Gossiper.instance.addLocalApplicationState(ApplicationState.INTERNAL_IP, StorageService.instance.valueFactory.internalIP(localPrivateAddress));
        Gossiper.instance.register(new ReconnectableSnitchHelper(this, ec2region, true));
    }
}
```

**Comment by DivsDibs** [comment]
`curl  http://169.254.169.254/latest/meta-data/local-ipv4l-ipv4` executed from the ec2-node is giving the correct pvt ip address, which is same as being used in configuration file.

**Highlighting** The 1st node starts fine. The 2nd node and 3rd node fails, irrespective they are seed node or not.

**Comment by vroyer** [comment]
The next release 6.2.3.8 will include a fix for that.

**Comment by gouthamaenugu11** [comment]
Vroyer,
I could see 6.2.3.9 available now, I see issue(#236) is still open. Is that right?

**Comment by vroyer** [comment]
Closed with no feedback about the fix provided in 6.2.3.9
